promql/engine.go

```diff
@@ -1343,29 +1359,36 @@ func (ev *evaluator) eval(expr parser.Expr) (parser.Value, storage.Warnings) {
 
 // vectorSelector evaluates a *parser.VectorSelector expression.
 func (ev *evaluator) vectorSelector(node *parser.VectorSelector, ts int64) (Vector, storage.Warnings) {
-	ws, err := checkAndExpandSeriesSet(ev.ctx, node)
-	if err != nil {
-		ev.error(errWithWarnings{errors.Wrap(err, "expanding series"), ws})
-	}
-	vec := make(Vector, 0, len(node.Series))
 	it := storage.NewBuffer(durationMilliseconds(ev.lookbackDelta))
-	for i, s := range node.Series {
-		it.Reset(s.Iterator())
+	if node.Series == nil {
+		uexset := node.UnexpandedSeriesSet
+		var vec Vector // Output Vector
+		for uexset.Next() {
+			s := uexset.At()
+			it.Reset(s.Iterator())
 
-		t, v, ok := ev.vectorSelectorSingle(it, node, ts)
-		if ok {
-			vec = append(vec, Sample{
-				Metric: node.Series[i].Labels(),
-				Point:  Point{V: v, T: t},
-			})
-			ev.currentSamples++
-		}
+			t, v, ok := ev.vectorSelectorSingle(it, node, ts)
+			if ok {
+				vec = append(vec, Sample{
+					Metric: s.Labels(),
+					Point:  Point{V: v, T: t},
+				})
+				ev.currentSamples++
+			}
 
-		if ev.currentSamples >= ev.maxSamples {
-			ev.error(ErrTooManySamples(env))
+			if ev.currentSamples >= ev.maxSamples {
+				ev.error(ErrTooManySamples(env))
+			}
 		}
+		ws := uexset.Warnings()
+		err := uexset.Err()
+		if err != nil {
+			ev.error(errWithWarnings{errors.Wrap(err, "expanding series"), ws})
+		}
+		return vec, ws
 	}
-	return vec, ws
+	vec := make(Vector, 0, 0) // Empty Vector to avoid returning a nil Vector
```

Comment:
I'd expect returning nil to be fine, is it causing problems?

Reply:
We return empty Vectors right now, so didn't want to break that.

Reply:
I have been trying to solve [this failing test](https://github.com/prometheus/prometheus/blob/master/promql/testdata/functions.test#L297). I think the problem here is that, `rangeEval()` makes the function call inside the for loop and since this change uses `.Next()`, instead of for looping through `node.Series`, we aren't able to execute the desired logic every iteration inside the `rangeEval()` method.

Reply:
That makes sense.

Reply:
Haven't been able to think of a solution, so I was thinking of just using `checkAndExpandSerieSet` in this function. Does that work or should we aim to remove all usages of `checkAndExpandSerieSet`?

Reply:
The point of this is to eliminate this that function.